workflow/libero_flow_poc/libero_flow/scheduler.py

```diff
@@ -0,0 +1,217 @@
+import copy
+import json
+import logging
+from typing import Any, Dict
+import time
+
+import pika
+from pika.exceptions import (
+    ChannelClosed,
+    ConnectionClosed,
+    IncompatibleProtocolError
+)
+import requests
+
+from libero_flow.conf import (
+    ACTIVITY_API_URL,
+    WORKFLOW_API_URL,
+    SCHEDULED_ACTIVITY_EXCHANGE,
+    SCHEDULED_DECISION_EXCHANGE,
+    ACTIVITY_RESULT_QUEUE,
+    DECISION_RESULT_QUEUE,
+    SCHEDULED_ACTIVITY_QUEUE,
+    SCHEDULED_DECISION_QUEUE,
+    WORKFLOW_STARTER_QUEUE,
+)
+from libero_flow.flow_loader import FlowLoader
+from libero_flow.event_utils import (
+    get_base_message,
+    get_channel,
+    message_handler,
+    setup_exchanges_and_queues,
+    DELIVERY_MODE_PERSISTENT,
+)
+from libero_flow.session_store import get_session
+from libero_flow.state_utils import (
+    get_activity_state,
+    send_workflow_event,
+    update_activity_status,
+    update_workflow_status,
+    FAILED,
+    FINISHED,
+    WORKFLOW_ACTIVITY_SCHEDULED,
+    WORKFLOW_DECISION_SCHEDULED,
+)
+
+logger = logging.getLogger(__name__)
+fh = logging.FileHandler(f'{__name__}.log')
+logger.addHandler(fh)
+
+
+def schedule_activity(activity_id: str, workflow_id: str) -> None:
+    """create and send schedule activity message.
+
+    :param activity_id: str
+    :param workflow_id: str
+    :return:
+    """
+    with get_channel() as channel:
+        message = get_base_message()
+        message['aggregate']['service'] = 'flow-scheduler'
+        message['aggregate']['name'] = 'schedule-workflow-activity'
+        message['type'] = 'schedule-activity'
+        message['data'] = {'activity_id': activity_id}
+
+        channel.basic_publish(exchange=SCHEDULED_ACTIVITY_EXCHANGE,
+                              routing_key=SCHEDULED_ACTIVITY_QUEUE,
```

Comment:
Am I getting RabbitMQ exchanges wrong or are all the `routing_key`s not needed? It should suffice to publish to the exchange which will fan out to its (only?) queue.

Reply:
Yes from my understanding, you should be able to publish to the exchange.

Though the `routing_key` is a required field:

`basic_publish(exchange, routing_key, body, properties=None, mandatory=False, immediate=False)`

I will investigate further for future use cases, but in this instance, removing the `routing_key` raises:

`TypeError: basic_publish() missing 1 required positional argument: 'routing_key'`

Reply:
The `fanout` exchange should ignore routing keys, so there should be some conventional value like `*` to pass in

Reply:
👍  I will investigate further.

Reply:
Sorted. The old empty string comes to the rescue.